File: LeetCode/cpp/792.cpp
class Solution {
public:
    int numMatchingSubseq(string S, vector<string> &words) {
        unordered_map<char, vector<int>> sub;
        int total = 0;
        for (int i = 0; i < S.size(); ++i) {
            if (sub.find(S[i]) == sub.end())
                sub[S[i]] = vector<int>();
            sub[S[i]].push_back(i);
        }
        for (int i = 0; i < words.size(); ++i) {
            auto &word = words[i];
            int j = 0, current_sub = -1;
            bool exist = true;
            while (j < word.size()) {
                auto c = word[j];
                if (sub[c].size() == 0) {
                    exist = false;
                    break;
                }
                int x = 0, y = sub[c].size(), mid = 0;
                while (x < y) {
                    mid = x + (y - x) / 2;
                    if (sub[c][mid] > current_sub)
                        y = mid;
                    else if (sub[c][mid] <= current_sub)
                        x = mid + 1;
                }
                if (x == sub[c].size()) {
                    exist = false;
                    break;
                } else
                    current_sub = sub[c][y];
                ++j;
            }
            if (exist)
                ++total;
        }
        return total;
    }
};

File: LeetCode/python3/792.py
class Solution:
    def numMatchingSubseq(self, S: 'str', words: 'List[str]') -> 'int':
        str = {}
        for s in range(len(S)):
            if S[s] not in str:
                str[S[s]] = []
            str[S[s]].append(s)
        total = 0
        for word in words:
            sub, exist = -1, True
            for w in word:
                if w not in str:
                    exist = False
                    break
                x, y, mid = 0, len(str[w]), 0
                while x < y:
                    mid = x + (y - x) // 2
                    if str[w][mid] > sub:
                        y = mid
                    else:
                        x = mid + 1
                if x == len(str[w]):
                    exist = False
                    break
                sub = str[w][y]
            if exist:
                total += 1
        return total